File: backend/app/services/quote_service.py
"""Сервис калькулятора: версионирование расчётов, строки, пересчёт, статусы.

Старые/непустые версии и принятые расчёты — read-only. Корректировка —
новая версия (clone), не редактирование (CLAUDE.md §11)."""

from __future__ import annotations

import uuid
from decimal import Decimal

from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import AppError, ErrorCode, NotFoundError
from app.models.enums import AuditAction, QuoteStatus
from app.models.quote import Quote, QuoteLine
from app.repositories.quote_repo import QuoteRepository
from app.schemas.quote import QuoteCreate, QuoteLineInput
from app.services import quote_calc as qc
from app.services.audit_service import AuditService


class QuoteService:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session
        self.repo = QuoteRepository(session)
        self.audit = AuditService(session)

    # Машина состояний расчёта. Вперёд из draft/sent — можно (включая прямой
    # accept). Принятый/отклонённый ТКП — ТЕРМИНАЛЬНЫЙ и неизменяемый: переходов
    # из него нет, правка только через новую версию (clone). CLAUDE.md §11.
    _ALLOWED_TRANSITIONS: dict[QuoteStatus, set[QuoteStatus]] = {
        QuoteStatus.DRAFT: {QuoteStatus.SENT, QuoteStatus.ACCEPTED, QuoteStatus.REJECTED},
        QuoteStatus.SENT: {QuoteStatus.ACCEPTED, QuoteStatus.REJECTED, QuoteStatus.DRAFT},
        QuoteStatus.ACCEPTED: set(),
        QuoteStatus.REJECTED: set(),
    }

    async def list_for_project(self, project_id: uuid.UUID) -> list[Quote]:
        return list(await self.repo.list_for_project(project_id))

    async def get(self, quote_id: uuid.UUID, project_id: uuid.UUID | None = None) -> Quote:
        quote = await self.repo.get_with_lines(quote_id, project_id)
        if quote is None:
            raise NotFoundError("Расчёт не найден")
        return quote

    @staticmethod
    def _assert_editable(quote: Quote) -> None:
        if quote.status != QuoteStatus.DRAFT:
            raise AppError(
                "Расчёт не в статусе «черновик» — редактирование запрещено "
                "(создайте новую версию)",
                code=ErrorCode.QUOTE_LOCKED,
                status_code=409,
            )

    async def create(
        self, project_id: uuid.UUID, data: QuoteCreate, actor_id: uuid.UUID
    ) -> Quote:
        version = await self.repo.max_version(project_id) + 1
        quote = await self.repo.create(
            project_id=project_id,
            version=version,
            title=data.title,
            status=QuoteStatus.DRAFT,
            currency_rates_snapshot=data.currency_rates,
            currency_buffer_pct=data.currency_buffer_pct,
            totals={},
        )
        await self.audit.record(
            entity="Quote",
            entity_id=quote.id,
            action=AuditAction.CREATE,
            actor_id=actor_id,
            diff={"version": version},
        )
        await self.session.flush()
        return await self.get(quote.id)

    def _line_to_calc_input(self, line: QuoteLine) -> qc.LineCalcInput:
        return qc.LineCalcInput(
            kind=line.kind,
            qty=line.qty,
            unit_price=line.unit_price,
            unit_cost=line.unit_cost,
            currency=line.currency,
            licensing_model=line.licensing_model,
            term_months=line.term_months,
            support_pct=line.support_pct,
            partner_discount_pct=line.partner_discount_pct,
            client_discount_pct=line.client_discount_pct,
        )

    async def recompute(self, quote: Quote) -> Quote:
        """Пересчитать все строки и итоги; сохранить результаты."""
        rates = {k: Decimal(str(v)) for k, v in (quote.currency_rates_snapshot or {}).items()}
        buffer = quote.currency_buffer_pct or Decimal("0")
        computed: list[tuple[qc.LineCalcInput, qc.LineCalcResult]] = []
        for line in quote.lines:
            calc_in = self._line_to_calc_input(line)
            try:
                res = qc.compute_line(calc_in, rates=rates, buffer_pct=buffer)
            except qc.RateNotFound as exc:
                raise AppError(
                    str(exc), code=ErrorCode.RATE_NOT_FOUND, status_code=422
                ) from exc
            line.cost_amount = res.cost_amount
            line.sell_amount = res.sell_amount
            line.margin = res.margin
            computed.append((calc_in, res))
        quote.totals = qc.compute_totals(computed).as_dict()
        await self.session.flush()
        return quote

    async def add_line(
        self, quote_id: uuid.UUID, data: QuoteLineInput, project_id: uuid.UUID | None = None
    ) -> Quote:
        quote = await self.get(quote_id, project_id)
        self._assert_editable(quote)
        line = QuoteLine(quote_id=quote.id, **data.model_dump())
        self.session.add(line)
        await self.session.flush()
        await self.session.refresh(quote, ["lines"])
        return await self.recompute(quote)

    async def update_line(
        self,
        quote_id: uuid.UUID,
        line_id: uuid.UUID,
        data: QuoteLineInput,
        project_id: uuid.UUID | None = None,
    ) -> Quote:
        quote = await self.get(quote_id, project_id)
        self._assert_editable(quote)
        line = next((ln for ln in quote.lines if ln.id == line_id), None)
        if line is None:
            raise NotFoundError("Строка расчёта не найдена")
        for field, value in data.model_dump().items():
            setattr(line, field, value)
        await self.session.flush()
        return await self.recompute(quote)

    async def delete_line(
        self, quote_id: uuid.UUID, line_id: uuid.UUID, project_id: uuid.UUID | None = None
    ) -> Quote:
        quote = await self.get(quote_id, project_id)
        self._assert_editable(quote)
        line = next((ln for ln in quote.lines if ln.id == line_id), None)
        if line is None:
            raise NotFoundError("Строка расчёта не найдена")
        await self.session.delete(line)
        await self.session.flush()
        await self.session.refresh(quote, ["lines"])
        return await self.recompute(quote)

    async def set_status(
        self,
        quote_id: uuid.UUID,
        status: QuoteStatus,
        actor_id: uuid.UUID,
        project_id: uuid.UUID | None = None,
    ) -> Quote:
        quote = await self.get(quote_id, project_id)
        if status != quote.status and status not in self._ALLOWED_TRANSITIONS[quote.status]:
            raise AppError(
                f"Недопустимый переход статуса расчёта: «{quote.status}» → «{status}». "
                "Принятый/отклонённый расчёт неизменяем — создайте новую версию.",
                code=ErrorCode.QUOTE_LOCKED,
                status_code=409,
            )
        quote.status = status
        await self.audit.record(
            entity="Quote",
            entity_id=quote.id,
            action=AuditAction.UPDATE,
            actor_id=actor_id,
            diff={"status": str(status)},
        )
        await self.session.flush()
        return quote

    async def clone_new_version(
        self, quote_id: uuid.UUID, actor_id: uuid.UUID, project_id: uuid.UUID | None = None
    ) -> Quote:
        """Создать новую черновую версию на основе существующей (копия строк)."""
        src = await self.get(quote_id, project_id)
        version = await self.repo.max_version(src.project_id) + 1
        new = await self.repo.create(
            project_id=src.project_id,
            version=version,
            title=src.title,
            status=QuoteStatus.DRAFT,
            currency_rates_snapshot=dict(src.currency_rates_snapshot or {}),
            currency_buffer_pct=src.currency_buffer_pct,
            totals={},
        )
        for line in src.lines:
            self.session.add(
                QuoteLine(
                    quote_id=new.id,
                    kind=line.kind,
                    name=line.name,
                    product_id=line.product_id,
                    licensing_model=line.licensing_model,
                    metric=line.metric,
                    currency=line.currency,
                    qty=line.qty,
                    unit_price=line.unit_price,
                    unit_cost=line.unit_cost,
                    term_months=line.term_months,
                    support_pct=line.support_pct,
                    partner_discount_pct=line.partner_discount_pct,
                    client_discount_pct=line.client_discount_pct,
                    note=line.note,
                )
            )
        await self.session.flush()
        new = await self.get(new.id)
        await self.audit.record(
            entity="Quote",
            entity_id=new.id,
            action=AuditAction.CREATE,
            actor_id=actor_id,
            diff={"cloned_from": str(src.id), "version": version},
        )
        return await self.recompute(new)

File: backend-dotnet/src/Veha.Api/Services/QuoteService.cs
using System.Globalization;
using Microsoft.EntityFrameworkCore;
using Veha.Api.Dtos;
using Veha.Domain.Common;
using Veha.Domain.Entities;
using Veha.Domain.Enums;
using Veha.Domain.Rules;
using Veha.Infrastructure;

namespace Veha.Api.Services;

/// <summary>Калькулятор ТКП: версии, строки, пересчёт, статусы, клонирование.
/// Порт services/quote_service.py + quote_repo.py. Правка — только черновик; принятый/
/// отклонённый — терминальный (правка через новую версию).</summary>
public class QuoteService(VehaDbContext db, AuditService audit)
{
    // Машина состояний: accepted/rejected — терминальные.
    private static readonly Dictionary<QuoteStatus, QuoteStatus[]> Allowed = new()
    {
        [QuoteStatus.Draft] = [QuoteStatus.Sent, QuoteStatus.Accepted, QuoteStatus.Rejected],
        [QuoteStatus.Sent] = [QuoteStatus.Accepted, QuoteStatus.Rejected, QuoteStatus.Draft],
        [QuoteStatus.Accepted] = [],
        [QuoteStatus.Rejected] = [],
    };

    public async Task<List<QuoteOutDto>> ListForProjectAsync(Guid projectId, CancellationToken ct)
        => (await db.Quotes.Where(q => q.ProjectId == projectId).OrderByDescending(q => q.Version).ToListAsync(ct))
            .Select(ToOut).ToList();

    public async Task<Quote> GetEntityAsync(Guid quoteId, Guid? projectId, CancellationToken ct)
    {
        var q = db.Quotes.Include(x => x.Lines).Where(x => x.Id == quoteId);
        if (projectId is not null) q = q.Where(x => x.ProjectId == projectId);
        return await q.FirstOrDefaultAsync(ct) ?? throw new NotFoundException("Расчёт не найден");
    }

    public async Task<QuoteDetailDto> GetAsync(Guid quoteId, Guid? projectId, CancellationToken ct)
        => ToDetail(await GetEntityAsync(quoteId, projectId, ct));

    private static void AssertEditable(Quote quote)
    {
        if (quote.Status != QuoteStatus.Draft)
            throw new QuoteLockedException(
                "Расчёт не в статусе «черновик» — редактирование запрещено (создайте новую версию)");
    }

    private async Task<int> MaxVersionAsync(Guid projectId, CancellationToken ct)
        => (await db.Quotes.IgnoreQueryFilters().Where(q => q.ProjectId == projectId)
                .Select(q => q.Version).ToListAsync(ct)).DefaultIfEmpty(0).Max();

    public async Task<QuoteDetailDto> CreateAsync(Guid projectId, QuoteCreateDto dto, Guid actorId, CancellationToken ct)
    {
        var version = await MaxVersionAsync(projectId, ct) + 1;
        var quote = new Quote
        {
            ProjectId = projectId, Version = version, Title = dto.Title, Status = QuoteStatus.Draft,
            CurrencyRatesSnapshot = dto.CurrencyRates, CurrencyBufferPct = dto.CurrencyBufferPct, Totals = new(),
        };
        db.Quotes.Add(quote);
        await audit.RecordAsync("Quote", quote.Id, AuditAction.Create, actorId, new { version }, ct);
        return await GetAsync(quote.Id, null, ct);
    }

    private async Task RecomputeAsync(Quote quote, CancellationToken ct)
    {
        var rates = quote.CurrencyRatesSnapshot.ToDictionary(
            k => k.Key.ToUpperInvariant(), v => decimal.Parse(v.Value, CultureInfo.InvariantCulture));
        var computed = new List<(QuoteCalc.LineCalcInput, QuoteCalc.LineCalcResult)>();
        foreach (var line in quote.Lines)
        {
            var input = ToCalcInput(line);
            var res = QuoteCalc.ComputeLine(input, rates, quote.CurrencyBufferPct);   // может бросить RateNotFound (422)
            line.CostAmount = res.CostAmount;
            line.SellAmount = res.SellAmount;
            line.Margin = res.Margin;
            computed.Add((input, res));
        }
        quote.Totals = QuoteCalc.ComputeTotals(computed);
        await db.SaveChangesAsync(ct);
    }

    public async Task<QuoteDetailDto> AddLineAsync(Guid quoteId, QuoteLineInputDto dto, Guid? projectId, CancellationToken ct)
    {
        var quote = await GetEntityAsync(quoteId, projectId, ct);
        AssertEditable(quote);
        db.QuoteLines.Add(FromInput(quote.Id, dto));
        await db.SaveChangesAsync(ct);
        var fresh = await GetEntityAsync(quoteId, projectId, ct);
        await RecomputeAsync(fresh, ct);
        return ToDetail(fresh);
    }

    public async Task<QuoteDetailDto> UpdateLineAsync(Guid quoteId, Guid lineId, QuoteLineInputDto dto, Guid? projectId, CancellationToken ct)
    {
        var quote = await GetEntityAsync(quoteId, projectId, ct);
        AssertEditable(quote);
        var line = quote.Lines.FirstOrDefault(l => l.Id == lineId)
                   ?? throw new NotFoundException("Строка расчёта не найдена");
        ApplyInput(line, dto);
        await RecomputeAsync(quote, ct);
        return ToDetail(quote);
    }

    public async Task<QuoteDetailDto> DeleteLineAsync(Guid quoteId, Guid lineId, Guid? projectId, CancellationToken ct)
    {
        var quote = await GetEntityAsync(quoteId, projectId, ct);
        AssertEditable(quote);
        var line = quote.Lines.FirstOrDefault(l => l.Id == lineId)
                   ?? throw new NotFoundException("Строка расчёта не найдена");
        db.QuoteLines.Remove(line);
        await db.SaveChangesAsync(ct);
        var fresh = await GetEntityAsync(quoteId, projectId, ct);
        await RecomputeAsync(fresh, ct);
        return ToDetail(fresh);
    }

    public async Task<QuoteOutDto> SetStatusAsync(Guid quoteId, QuoteStatus status, Guid actorId, Guid? projectId, CancellationToken ct)
    {
        var quote = await GetEntityAsync(quoteId, projectId, ct);
        if (status != quote.Status && !Allowed[quote.Status].Contains(status))
            throw new QuoteLockedException(
                $"Недопустимый переход статуса расчёта: «{quote.Status}» → «{status}». " +
                "Принятый/отклонённый расчёт неизменяем — создайте новую версию.");
        quote.Status = status;
        await audit.RecordAsync("Quote", quote.Id, AuditAction.Update, actorId, new { status }, ct);
        return ToOut(quote);
    }

    public async Task<QuoteDetailDto> CloneAsync(Guid quoteId, Guid actorId, Guid? projectId, CancellationToken ct)
    {
        var src = await GetEntityAsync(quoteId, projectId, ct);
        var version = await MaxVersionAsync(src.ProjectId, ct) + 1;
        var newQuote = new Quote
        {
            ProjectId = src.ProjectId, Version = version, Title = src.Title, Status = QuoteStatus.Draft,
            CurrencyRatesSnapshot = new(src.CurrencyRatesSnapshot), CurrencyBufferPct = src.CurrencyBufferPct, Totals = new(),
        };
        db.Quotes.Add(newQuote);
        foreach (var line in src.Lines)
            db.QuoteLines.Add(new QuoteLine
            {
                QuoteId = newQuote.Id, Kind = line.Kind, Name = line.Name, ProductId = line.ProductId,
                LicensingModel = line.LicensingModel, Metric = line.Metric, Currency = line.Currency,
                Qty = line.Qty, UnitPrice = line.UnitPrice, UnitCost = line.UnitCost, TermMonths = line.TermMonths,
                SupportPct = line.SupportPct, PartnerDiscountPct = line.PartnerDiscountPct,
                ClientDiscountPct = line.ClientDiscountPct, Note = line.Note,
            });
        await db.SaveChangesAsync(ct);
        var fresh = await GetEntityAsync(newQuote.Id, null, ct);
        await audit.RecordAsync("Quote", newQuote.Id, AuditAction.Create, actorId,
            new { cloned_from = src.Id.ToString(), version }, ct);
        await RecomputeAsync(fresh, ct);
        return ToDetail(fresh);
    }

    // --- Мапперы ---
    private static QuoteCalc.LineCalcInput ToCalcInput(QuoteLine l) => new()
    {
        Kind = l.Kind, Qty = l.Qty, UnitPrice = l.UnitPrice, UnitCost = l.UnitCost, Currency = l.Currency,
        LicensingModel = l.LicensingModel, TermMonths = l.TermMonths, SupportPct = l.SupportPct,
        PartnerDiscountPct = l.PartnerDiscountPct, ClientDiscountPct = l.ClientDiscountPct,
    };

    private static QuoteLine FromInput(Guid quoteId, QuoteLineInputDto d) => new()
    {
        QuoteId = quoteId, Kind = d.Kind, Name = d.Name, ProductId = d.ProductId, LicensingModel = d.LicensingModel,
        Metric = d.Metric, Currency = d.Currency, Qty = d.Qty, UnitPrice = d.UnitPrice, UnitCost = d.UnitCost,
        TermMonths = d.TermMonths, SupportPct = d.SupportPct, PartnerDiscountPct = d.PartnerDiscountPct,
        ClientDiscountPct = d.ClientDiscountPct, Note = d.Note,
    };

    private static void ApplyInput(QuoteLine l, QuoteLineInputDto d)
    {
        l.Kind = d.Kind; l.Name = d.Name; l.ProductId = d.ProductId; l.LicensingModel = d.LicensingModel;
        l.Metric = d.Metric; l.Currency = d.Currency; l.Qty = d.Qty; l.UnitPrice = d.UnitPrice; l.UnitCost = d.UnitCost;
        l.TermMonths = d.TermMonths; l.SupportPct = d.SupportPct; l.PartnerDiscountPct = d.PartnerDiscountPct;
        l.ClientDiscountPct = d.ClientDiscountPct; l.Note = d.Note;
    }

    private static void FillOut(QuoteOutDto o, Quote q)
    {
        o.Id = q.Id; o.ProjectId = q.ProjectId; o.Version = q.Version; o.Title = q.Title; o.Status = q.Status;
        o.CurrencyRatesSnapshot = new(q.CurrencyRatesSnapshot); o.CurrencyBufferPct = q.CurrencyBufferPct;
        o.Totals = new(q.Totals); o.CreatedAt = q.CreatedAt;
    }

    public static QuoteOutDto ToOut(Quote q) { var o = new QuoteOutDto(); FillOut(o, q); return o; }

    public static QuoteDetailDto ToDetail(Quote q)
    {
        var d = new QuoteDetailDto { Lines = q.Lines.Select(ToLineOut).ToList() };
        FillOut(d, q);
        return d;
    }

    private static QuoteLineOutDto ToLineOut(QuoteLine l) => new(
        l.Id, l.Kind, l.Name, l.ProductId, l.LicensingModel, l.Metric, l.Currency, l.Qty, l.UnitPrice, l.UnitCost,
        l.TermMonths, l.SupportPct, l.PartnerDiscountPct, l.ClientDiscountPct, l.CostAmount, l.SellAmount, l.Margin, l.Note);
}
